csrc/optimization/move_split_cat.cpp

```diff
@@ -15,11 +15,96 @@
 #include <ir/interface_nodes.h>
 #include <ir/internal_base_nodes.h>
 #include <ir/utils.h>
+#include <ops/alias.h>
 
 namespace nvfuser::optimization {
 
 namespace {
 
+// Returns true when Exprs in the frontier can be horizontally merged and
+// applied on the unsplit tensor.
+bool horizontallyMergeable(
+    const std::vector<Expr*>& frontier,
+    int64_t& split_axis) {
+  NVF_ERROR(!frontier.empty());
+
+  // Check all Exprs in `frontier`
+  // 1. have the same op type and attributes,
+  // 2. transform IDs in the same way, and
+  // 3. don't resize the split axis.
+
+  if (std::adjacent_find(
+          frontier.begin(), frontier.end(), [](Expr* lhs, Expr* rhs) {
+            return !lhs->sameOp(rhs);
+          }) != frontier.end()) {
+    return false;
+  }
+
+  if (auto* set = dynamic_cast<LoadStoreOp*>(frontier[0])) {
```

Comment:
qq: what happens when we encounter a reduction on the split_axis?

Reply:
This function will skip the op and return false, which is a conservative but correct answer. Or are you saying LoadStoreOp can reduce an axis? (I wasn't aware of that).

Reply:
Ha, sorry I was blind and didn't see there was a `return false` in the end.

you are right that `LoadStoreOp*` here should be pretty safe.